Add threshold continuous uniform distribution button

Sonje opened the pull request:
### Changes:
- added a button that distributes threshold thumbs within min and max measurements taken for the given session details or calendar page
- added error when min and max are the same value (reused threshold configurator error, it gets cleared after 3 seconds of button interaction)
- created a reusable threshold button, so it's easier to apply styles for different designs
- refactored reset button to utilise the default threshold button
- the changes as per ticket are only applied to desktop design
- the distribution also works with negative values
- fixed incorrect typing for fixed stream `sessionId` property - we receive it as a number from API, though it was typed incorrectly as a string everywhere (and comparisons were frequently forced to use string conversion unnecessarily

### Error:
![image](https://github.com/user-attachments/assets/4f1a76da-9b7f-4cbc-8e90-6011062aefdf)

### 0-1 example (users can't type the fractions but we have to show them for the proper thumb placement)

![Zrzut ekranu 2024-08-19 183618](https://github.com/user-attachments/assets/a79b6c2a-4939-4ab9-b753-3ad04e01a685)

### Calendar preview

https://github.com/user-attachments/assets/23b8ccad-a4b6-467d-b1dd-b73e95b9a38b

### Negative values example

https://github.com/user-attachments/assets/b3d0e2e1-f0a8-4c9d-8413-cacd4015d490

Sonje commented:
![](https://github.trello.services/images/mini-trello-icon.png) [Session Details: Adjust Breakpoint Values to Measurements](https://trello.com/c/dhlzzfsK/1836-session-details-adjust-breakpoint-values-to-measurements)

JoannaThelema commented:
This marker is available in the New York location. 
![Screenshot 2024-08-21 at 10 17 42](https://github.com/user-attachments/assets/7fef8418-2b4b-4442-8fb4-94ec475fe1d5)

1. Sometimes the graph is cut off, I suspect it's because we're rounding displayed values and we may have 0.8 on the graph.
2. I see that we're missing one threshold button (between orange and red) 
![Screenshot 2024-08-21 at 10 06 44](https://github.com/user-attachments/assets/7615f3e5-a91f-4098-b0fa-2a1099f29bd9)

3. When I reload the page, values are back to default (not how it works on current prod)


##### The same point from prod:
<img width="861" alt="Screenshot 2024-08-21 at 10 33 30" src="https://github.com/user-attachments/assets/7106c522-0880-451d-934a-f1052d876ba2">

JoannaThelema commented:
Since we don't have a separate ticket for mobile devices and we're missing reset button only on calendar page I would do it in this pr 💪

JoannaThelema commented on app/javascript/react/components/ThresholdConfigurator/ThresholdButtons/UniformDistributionButton.tsx:
```diff
@@ -0,0 +1,90 @@
+      mobileStream.status === StatusEnum.Fulfilled &&
+      mobileStream.data.id === sessionId
+    ) {
+      min = Math.round(mobileStream.minMeasurementValue!);
```
I think we need to round the min value always to the lower one e.g. `Math.flor`

JoannaThelema commented on app/javascript/react/components/ThresholdConfigurator/ThresholdButtons/UniformDistributionButton.tsx:
```diff
@@ -0,0 +1,90 @@
+      mobileStream.data.id === sessionId
+    ) {
+      min = Math.round(mobileStream.minMeasurementValue!);
+      max = Math.round(mobileStream.maxMeasurementValue!);
```
I think we need to round the max value always to the higher one e.g. `Math.ceil`

JoannaThelema commented on app/javascript/react/components/ThresholdConfigurator/ThresholdButtons/UniformDistributionButton.tsx:
```diff
@@ -0,0 +1,90 @@
+      fixedStream.status === StatusEnum.Fulfilled &&
+      fixedStream.data.stream.sessionId === sessionId
+    ) {
+      min = Math.round(fixedStream.minMeasurementValue!);
+      max = Math.round(fixedStream.maxMeasurementValue!);
```
Same for fixed stream

JoannaThelema reviewed (commented):
The code looks good 🙌 
I found one additional bug: loading thresholds from URL is broken for map and modal view 😅

JoannaThelema commented on app/javascript/react/store/thresholdSlice.ts:
```diff
@@ -26,10 +26,10 @@ export interface ThresholdState {
-    middle: 0,
-    high: 0,
-    max: 0,
+    low: 9,
+    middle: 35,
+    high: 55,
+    max: 150,
```
Why do we need to change defaults? I think this is creating the bug I mentioned earlier. If we want to keep it like that we need to improve conditions for loading thresholds
```
  useEffect(() => {
    if (!isFirstRenderForThresholds.current) {
      dispatch(resetUserThresholds());
    }
    // #DirtyButWorks :nervous-laugh: -> refactor when moving thresholds to url
    if (defaultThresholds.max !== 0) {
      isFirstRenderForThresholds.current = false;
    }
  }, [defaultThresholds]);

  useEffect(() => {
    if (isFirstRenderForThresholds.current) {
      dispatch(setUserThresholdValues(initialThresholds));
    }
    // #DirtyButWorks :nervous-laugh: -> refactor when moving thresholds to url
    if (initialThresholds.max === 0) {
      isFirstRenderForThresholds.current = false;
    }
  }, [initialThresholds]);

Sonje commented:
@JoannaThelema thanks so much for looking into the issues.
CR changes:
- changed threshold distribution logic from discrete uniform distribution to continuous uniform distribution as we have in production right now,
- fixed issue when slider would sometimes cover the max value (it was a threshold slider bug),
- in the meantime decoupled the logic and calculations from click button
- adjusted the code and styling so the new button can be added to mobile calendar page
- added the button to mobile calendar page with a placeholder button/s - the design is used in mobile map legend so had to adjust it accordingly - added the comment to remove them after adding theme switch functionality

JoannaThelema commented on app/javascript/react/components/ThresholdConfigurator/ThresholdSlider.tsx:
```diff
@@ -190,7 +190,23 @@ const ThresholdSlider: React.FC<ThresholdSliderProps> = ({
       onFocus={() => handleInputFocus(thresholdKey)}
       onBlur={() => handleInputBlur(thresholdKey, inputValue)}
-      style={positionStyle}
+      style={{
+        ...positionStyle,
+        zIndex: 10,
+        marginLeft:
+          thresholdKey === "min" ? "-1px" : value === max ? "0px" : "-15px",
+        left:
+          thresholdKey === "max"
+            ? "auto"
+            : `${calculateThumbPosition(
+                value,
+                min,
+                max,
+                sliderWidth,
+                isMobile
+              )}px`,
+        right: thresholdKey === "max" ? "-1px" : "auto",
+      }}
```
🚀

JoannaThelema reviewed (approved):
Great job! 🙌 Let's merge 😎
